dataset/bucket.cpp

```diff
@@ -188,4 +199,71 @@ void append(const VariableView &var0, const VariableConstView &var1) {
     var0.replace_model(combine<Dataset>(var0, var1));
 }
 
+namespace histogram_detail {
+template <class Out, class Coord, class Weight, class Edge>
+using args = std::tuple<span<Out>, span<const Coord>, span<const Weight>,
+                        span<const Edge>>;
+}
+
+Variable histogram(const VariableConstView &data,
+                   const VariableConstView &binEdges) {
+  using namespace scipp::core;
+  using namespace histogram_detail;
+  auto hist_dim = binEdges.dims().inner();
+  const auto &[indices, dim, buffer] = data.constituents<bucket<DataArray>>();
+  if (!buffer.masks().empty())
+    throw std::runtime_error("Masked data cannot be histogrammed yet.");
+  VariableConstView spans(indices);
+  Variable merged;
+  if (indices.dims().contains(hist_dim)) {
+    const auto size = indices.dims()[hist_dim];
+    const auto [begin, end] = unzip(indices);
+    // Only contiguous ranges along histogramming dim supported at this point.
+    core::expect::equals(begin.slice({hist_dim, 1, size}),
+                         end.slice({hist_dim, 0, size - 1}));
+    merged = zip(begin.slice({hist_dim, 0}), end.slice({hist_dim, size - 1}));
+    spans = VariableConstView(merged);
+  }
+  return variable::transform_subspan<std::tuple<
+      args<double, double, double, double>, args<double, float, double, double>,
+      args<double, float, double, float>, args<double, double, float, double>>>(
+      buffer.dtype(), hist_dim, binEdges.dims()[hist_dim] - 1,
+      subspan_view(buffer.coords()[hist_dim], dim, spans),
+      subspan_view(buffer.data(), dim, spans), binEdges, element::histogram);
+}
+
+Variable map(const DataArrayConstView &function, const VariableConstView &x,
+             Dim hist_dim) {
+  if (hist_dim == Dim::Invalid)
+    hist_dim = edge_dimension(function);
+  const auto mask = irreducible_mask(function.masks(), hist_dim);
+  Variable masked;
+  if (mask)
+    masked = function.data() * ~mask;
+  const auto &[indices, dim, buffer] = x.constituents<bucket<DataArray>>();
+  // Note the current inefficiency here: Output buffer is created with full
+  // size, even of `x` is a slice and only subsections of the buffer are needed.
+  auto out = variable::variableFactory().create(function.dtype(), buffer.dims(),
+                                                function.hasVariances());
+  transform_in_place(subspan_view(out, dim, indices),
+                     subspan_view(buffer.coords()[hist_dim], dim, indices),
+                     subspan_view(function.coords()[hist_dim], hist_dim),
+                     subspan_view(mask ? masked : function.data(), hist_dim),
+                     core::element::event::map_in_place);
+  return Variable{std::make_unique<variable::DataModel<bucket<Variable>>>(
+      indices, dim, std::move(out))};
+}
+
+void scale(const DataArrayView &data, const DataArrayConstView &histogram) {
```

Comment:
It should be noted that you *can* just use `data *= histogram`, *provided that bucketing matches binning*. `scale` is more sophisticated and can deal with a histogram of higher resolution, looking up the coord value of each event in each bucket.